examples/elasto_plastic.jl

```diff
@@ -0,0 +1,300 @@
+using JuAFEM
+using Tensors
+using KrylovMethods
+using TimerOutputs
+using Statistics: mean
+import ProgressMeter
+const ∇ = Tensors.gradient
+
+struct Material{T}
+    μ::Float64		#Lamé constant
+    λ::Float64		#Lamé constant
+    τ₀::Float64		#Yield stress
+    H::Float64		#Hardening
+end
+
+struct MaterialState{T}
+	S::SymmetricTensor{2,3,T,6} #2nd PK stress
+    ϵᵖ::Float64					#Plastic strain
+    F⁻¹::Tensor{2,3,T,9}		#Plastic deformation grad
+    ν::Tensor{2,3,T,9}			
+end
+
+function MaterialState()
+    return MaterialState(
+    				zero(SymmetricTensor{2,3,Float64,6}),
+    				0.0, 
+    	            one(Tensor{2,3,Float64,9}), 
+    	            zero(Tensor{2,3,Float64,9}),)
+end
+
+function _compute_2nd_PK(mp::Material, E, state::MaterialState)
+    μ, λ, τ₀, H = mp.μ, mp.λ, mp.τ₀, mp.H
+    
+    ϵᵖ = state.ϵᵖ
+    F⁻¹ = state.F⁻¹
+    ν = state.ν
+    dγ = 0 
+
+    C = 2E + one(E)
+
+    phi, Mdev, Stilde = ϕ(C, F⁻¹, ϵᵖ, μ, λ, τ₀, H)
+
+    if phi > 0 #Plastic step
+        dγ = 0.0
+        
+        #Setup newton varibles
+        TOL = 0.00001
+        newton_error, counter = TOL + 1, 0
+
+        # Define residual for semi-emplicit backward euler (assuming ⁿ⁺¹ν = ⁿν)
+        ϕ_residual(dγ) = ϕ(C, F⁻¹ - dγ*F⁻¹⋅ν, ϵᵖ + dγ, μ, λ, τ₀, H)[1]
+
+        while newton_error > TOL; counter +=1 
+            
+            res = ϕ_residual(dγ)
+
+            #Not able to take Automatic derivative of ϕ_residual for some reaseon
+            #Use numerical derivative for simplicity
+
+            #jac = ForwardDiff.derivative(ϕ_residual, dγ)
+            jac = (ϕ_residual(dγ + 0.00001) - ϕ_residual(dγ))/(0.00001)
```

Comment:
I am not able to use ` ForwardDiff.derivative` here because the `_compute_2nd_PK` is also callen with `Tensors.gradient`. Any way to fix this?

Reply:
Yes. You shouldn't use AD on an iterative solver, that will be crazy expensive. Instead, exploit that `dS/dE = inv(-dr/dS) * dr/dE` evaluated at the solved state where `r` is the residual function.

Reply:
I have looked at this for a while now, but I dont quite understand. My residual is `ϕ_residual` (it is a scalar), but it does not depend on SPK. how should I take `dr/dS`?

Reply:
I realised that  I should calculate `dF/dC` from the residual and use that to calculate `S`